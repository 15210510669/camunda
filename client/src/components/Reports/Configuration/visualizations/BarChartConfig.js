import React from 'react';
import {ColorPicker, Switch, LabeledInput} from 'components';
import ChartTargetInput from './subComponents/ChartTargetInput';

export default function BarChartConfig({onChange, report}) {
  const {
    combined,
    data: {configuration}
  } = report;

  return (
    <div className="BarChartConfig">
      {!combined && (
        <fieldset className="ColorSection">
          <legend>Select visualization color</legend>
          <ColorPicker
            selectedColor={configuration.color}
            onChange={color => onChange({color: {$set: color}})}
          />
        </fieldset>
      )}
      <fieldset className="axisConfig">
        <legend>Axis names</legend>
        <LabeledInput
          label="x-axis"
          type="text"
          value={configuration.xLabel}
          onChange={({target: {value}}) => onChange({xLabel: {$set: value}})}
        />
        <LabeledInput
          label="y-axis"
          type="text"
          value={configuration.yLabel}
          onChange={({target: {value}}) => onChange({yLabel: {$set: value}})}
        />
      </fieldset>
      <fieldset className="goalLine">
        <legend>
          <Switch
            checked={configuration.targetValue.active}
            onChange={({target: {checked}}) => onChange({targetValue: {active: {$set: checked}}})}
          />
          Goal
        </legend>
        <ChartTargetInput {...{onChange, report}} />
      </fieldset>
    </div>
  );
}
