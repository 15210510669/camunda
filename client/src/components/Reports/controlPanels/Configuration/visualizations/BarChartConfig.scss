.BarChartConfig {
  .axisConfig .Input {
    width: 100%;
  }
}
