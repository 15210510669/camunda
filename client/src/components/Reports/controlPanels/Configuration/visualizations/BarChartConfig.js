/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React from 'react';
import {ColorPicker, Switch, Input} from 'components';
import ChartTargetInput from './subComponents/ChartTargetInput';
import RelativeAbsoluteSelection from './subComponents/RelativeAbsoluteSelection';
import {isDurationReport} from 'services';
import './BarChartConfig.scss';
import {t} from 'translation';

export default function BarChartConfig({onChange, report}) {
  const {
    reportType,
    combined,
    data: {configuration, groupBy, distributedBy, visualization},
    result,
  } = report;

  const durationReport = isDurationReport(combined ? Object.values(result.data)[0] : report);
  const isMultiMeasure = combined ? false : result?.measures.length > 1;
  const isStackingPossible =
    !combined &&
    distributedBy.type !== 'none' &&
    groupBy.type !== 'none' &&
    ['barLine', 'bar'].includes(visualization);
  const isStacked = isStackingPossible && configuration.stackedBar;
  const isHorizontal = configuration.horizontalBar;

  return (
    <div className="BarChartConfig">
      {!combined &&
        !isMultiMeasure &&
        (distributedBy.type === 'none' ||
          (distributedBy.type === 'process' && groupBy.type === 'none')) && (
          <fieldset className="colorSection">
            <legend>{t('report.config.colorPicker.legend')}</legend>
            <ColorPicker
              selectedColor={configuration.color}
              onChange={(color) => onChange({color: {$set: color}})}
            />
          </fieldset>
        )}
      <fieldset>
        <legend>{t('report.config.tooltips.legend')}</legend>
        <RelativeAbsoluteSelection
          reportType={reportType}
          hideRelative={durationReport}
          absolute={configuration.alwaysShowAbsolute}
          relative={configuration.alwaysShowRelative}
          onChange={(type, value) => {
            if (type === 'absolute') {
              onChange({alwaysShowAbsolute: {$set: value}});
            } else {
              onChange({alwaysShowRelative: {$set: value}});
            }
          }}
        />
      </fieldset>
      {isStackingPossible && (
        <fieldset className="stackedBars">
          <legend>{t('report.config.stackedBars.legend')}</legend>
          <Switch
            checked={configuration.stackedBar}
            onChange={({target: {checked}}) => onChange({stackedBar: {$set: checked}})}
            label={t('report.config.stackedBars.enableStackedBars')}
          />
        </fieldset>
      )}
      <fieldset>
        <legend>{t('report.config.axisSettings.legend')}</legend>
        <Switch
          checked={configuration.logScale}
          onChange={({target: {checked}}) => onChange({logScale: {$set: checked}})}
          label={t('report.config.axisSettings.logScale')}
        />
        <label>{t('report.config.axisSettings.label')}</label>
        <Input
          placeholder={t('report.config.axisSettings.' + (isHorizontal ? 'yAxis' : 'xAxis'))}
          type="text"
          value={configuration.xLabel}
          onChange={({target: {value}}) => onChange({xLabel: {$set: value}})}
        />
        {!isMultiMeasure && (
          <Input
            placeholder={t('report.config.axisSettings.' + (isHorizontal ? 'xAxis' : 'yAxis'))}
            type="text"
            value={configuration.yLabel}
            onChange={({target: {value}}) => onChange({yLabel: {$set: value}})}
          />
        )}
      </fieldset>
      {!isMultiMeasure && !isStacked && (
        <fieldset>
          <legend>
            <Switch
              checked={configuration.targetValue.active}
              onChange={({target: {checked}}) => onChange({targetValue: {active: {$set: checked}}})}
              label={t('report.config.goal.legend')}
            />
          </legend>
          <ChartTargetInput {...{onChange, report}} />
        </fieldset>
      )}
    </div>
  );
}
