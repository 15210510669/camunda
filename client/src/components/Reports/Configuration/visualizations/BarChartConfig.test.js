import React from 'react';
import {shallow} from 'enzyme';
import BarChartConfig from './BarChartConfig';

const barReport = {
  combined: false,
  data: {visualization: 'bar', view: {property: 'frequency'}}
};

const props = {
  report: {combined: false}
};

const configuration = {
  showInstanceCount: false,
  color: '#1991c8',
  hideRelativeValue: false,
  hideAbsoluteValue: false,
  xLabel: '',
  yLabel: '',
  targetValue: {active: false}
};

it('it should display correct configuration for barchart', () => {
  const node = shallow(<BarChartConfig {...{report: barReport, configuration}} />);
  expect(node).toMatchSnapshot();
});

it('should reset to defaults when the property changes', () => {
  expect(
    BarChartConfig.onUpdate(
      {report: {data: {view: {property: 'prev'}}}},
      {report: {data: {view: {property: 'new'}}}}
    )
  ).toEqual({
    ...BarChartConfig.defaults(props),
    targetValue: null
  });
});

it('should reset to defaults when the entity changes', () => {
  expect(
    BarChartConfig.onUpdate(
      {report: {data: {view: {entity: 'prev'}}}},
      {report: {data: {view: {entity: 'new'}}}}
    )
  ).toEqual({
    ...BarChartConfig.defaults(props),
    targetValue: null
  });
});

it('should reset to defaults when visualization type changes', () => {
  expect(
    BarChartConfig.onUpdate(
      {type: 'prev', report: {data: {view: {entity: 'test'}}}},
      {type: 'new', report: {data: {view: {entity: 'test'}}}}
    )
  ).toEqual({
    ...BarChartConfig.defaults(props),
    targetValue: null
  });
});

it('should not reset to defaults when visualization type changes from line to bar or reverse', () => {
  expect(
    BarChartConfig.onUpdate(
      {type: 'bar', report: {data: {view: {entity: 'test'}}}},
      {type: 'line', report: {data: {view: {entity: 'test'}}}}
    )
  ).toEqual(undefined);
});

it('should reset to defaults when updating combined report type', () => {
  expect(
    BarChartConfig.onUpdate(
      {
        report: {
          combined: true
        }
      },
      {
        report: {
          combined: true
        },
        type: 'bar'
      }
    )
  ).toEqual(BarChartConfig.defaults({report: {combined: true}}));
});

it('should not display show instance count and color picker for combined reports', () => {
  const node = shallow(
    <BarChartConfig
      {...{
        report: {combined: true, result: {test: {data: {view: {property: 'frequency'}}}}},
        configuration
      }}
    />
  );

  expect(node.find('ShowInstanceCount')).not.toBePresent();
  expect(node.find('ColorPicker')).not.toBePresent();
});
