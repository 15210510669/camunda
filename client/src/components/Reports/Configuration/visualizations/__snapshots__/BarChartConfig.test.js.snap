// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should display correct configuration for barchart 1`] = `
<div
  className="BarChartConfig"
>
  <fieldset
    className="ColorSection"
  >
    <legend>
      Select visualization color
    </legend>
    <ColorPicker
      onChange={[Function]}
      selectedColor="#1991c8"
    />
  </fieldset>
  <fieldset
    className="axisConfig"
  >
    <legend>
      Axis names
    </legend>
    <LabeledInput
      label="x-axis"
      onChange={[Function]}
      type="text"
      value=""
    />
    <LabeledInput
      label="y-axis"
      onChange={[Function]}
      type="text"
      value=""
    />
  </fieldset>
  <fieldset
    className="goalLine"
  >
    <legend>
      <Switch
        checked={false}
        onChange={[Function]}
      />
      Goal
    </legend>
    <ChartTargetInput
      report={
        Object {
          "combined": false,
          "data": Object {
            "configuration": Object {
              "color": "#1991c8",
              "hideAbsoluteValue": false,
              "hideRelativeValue": false,
              "showInstanceCount": false,
              "targetValue": Object {
                "active": false,
              },
              "xLabel": "",
              "yLabel": "",
            },
            "view": Object {
              "property": "frequency",
            },
            "visualization": "bar",
          },
        }
      }
    />
  </fieldset>
</div>
`;
